pallets/ddc-staking/src/mock.rs

```diff
@@ -0,0 +1,244 @@
+	type UnixTime = Timestamp;
+	type WeightInfo = ();
+}
+
+pub(crate) type DdcStakingCall = crate::Call<Test>;
+pub(crate) type TestRuntimeCall = <Test as frame_system::Config>::Call;
+
+pub struct ExtBuilder {
+	has_edges: bool,
+	has_storages: bool,
+	stakes: BTreeMap<AccountId, Balance>,
+	edges: Vec<(AccountId, AccountId, Balance, ClusterId)>,
+	storages: Vec<(AccountId, AccountId, Balance, ClusterId)>,
+}
+
+impl Default for ExtBuilder {
+	fn default() -> Self {
+		Self {
+			has_edges: true,
+			has_storages: true,
+			stakes: Default::default(),
+			edges: Default::default(),
+			storages: Default::default(),
+		}
+	}
+}
+
+impl ExtBuilder {
+	pub fn has_edges(mut self, has: bool) -> Self {
+		self.has_edges = has;
+		self
+	}
+	pub fn has_storages(mut self, has: bool) -> Self {
+		self.has_storages = has;
+		self
+	}
+	pub fn set_stake(mut self, who: AccountId, stake: Balance) -> Self {
+		self.stakes.insert(who, stake);
+		self
+	}
+	pub fn add_edge(
+		mut self,
+		stash: AccountId,
+		controller: AccountId,
+		stake: Balance,
+		cluster: ClusterId,
+	) -> Self {
+		self.edges.push((stash, controller, stake, cluster));
+		self
+	}
+	pub fn add_storage(
+		mut self,
+		stash: AccountId,
+		controller: AccountId,
+		stake: Balance,
+		cluster: ClusterId,
+	) -> Self {
+		self.storages.push((stash, controller, stake, cluster));
+		self
+	}
+	fn build(self) -> TestExternalities {
+		sp_tracing::try_init_simple();
+		let mut storage = frame_system::GenesisConfig::default().build_storage::<Test>().unwrap();
+
+		let _ = pallet_balances::GenesisConfig::<Test> {
+			balances: vec![
+				(1, 100),
+				(2, 100),
+				(3, 100),
+				(4, 100),
+				// edge controllers
+				(10, 100),
+				(20, 100),
+				// storage controllers
+				(30, 100),
+				(40, 100),
+				// edge stashes
+				(11, 100),
+				(21, 100),
+				// storage stashes
+				(31, 100),
+				(41, 100),
+			],
+		}
+		.assimilate_storage(&mut storage);
+		let mut edges = vec![];
+		if self.has_edges {
+			edges = vec![
+				// (stash, controller, stake, cluster)
+				(11, 10, 100, 1),
+				(21, 20, 100, 1),
+			];
+		}
+		let mut storages = vec![];
+		if self.has_storages {
+			storages = vec![
+				// (stash, controller, stake, cluster)
+				(31, 30, 100, 1),
+				(41, 40, 100, 1),
+			];
+		}
+
+		let _ = pallet_ddc_staking::GenesisConfig::<Test> { edges, storages, ..Default::default() }
+			.assimilate_storage(&mut storage);
+
+		TestExternalities::new(storage)
+	}
+	pub fn build_and_execute(self, test: impl FnOnce() -> ()) {
+		sp_tracing::try_init_simple();
+		let mut ext = self.build();
+		ext.execute_with(test);
+		ext.execute_with(post_condition);
+	}
+}
+
+fn post_condition() {
+	check_ledgers();
+}
+
+fn check_ledgers() {
+	// check the ledger of all stakers.
+	Bonded::<Test>::iter().for_each(|(_, controller)| assert_ledger_consistent(controller))
+}
+
+fn assert_ledger_consistent(controller: AccountId) {
+	// ensures ledger.total == ledger.active + sum(ledger.unlocking).
```

Comment:
Do we need this comment?

Reply:
Yes, it may look like a code, but it is just a valid comment explaining the following stuff.